ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+         A nice by-effect is that this function doesn't mutate state in failure
+      *)
+      Option.iter (fun log -> undo_changes !log) log;
+      Error e
+
+  let submode obj a b =
+    match submode_try obj a b with
+    | Ok log ->
+      Option.iter !append_changes log;
+      Ok ()
+    | Error _ as e -> e
+
+  let zap_to_ceil_morphvar obj mv =
+    assert (submode obj (Amode (mupper obj mv)) (Amodevar mv) |> Result.is_ok);
+    mupper obj mv
+
+  let zap_to_ceil : type a l. a C.obj -> (a, l * allowed) mode -> a =
+   fun obj -> function
+    | Amode m -> m
+    | Amodevar mv -> zap_to_ceil_morphvar obj mv
+    | Amodemeet (a, mvs) ->
+      List.fold_left
+        (fun acc mv -> C.meet obj acc (zap_to_ceil_morphvar obj mv))
+        a mvs
+
+  let join (type a r) obj l =
+    let rec loop :
+        a ->
+        (a, allowed * disallowed) morphvar list ->
+        (a, allowed * r) mode list ->
+        (a, allowed * disallowed) mode =
+     fun a mvs ->
+      if C.le obj (C.max obj) a
+      then fun _ -> Amode (C.max obj)
+      else
+        function
+        | [] -> Amodejoin (a, mvs)
+        | mv :: xs -> (
+          match disallow_right mv with
+          | Amode b -> loop (C.join obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
+          | Amodejoin (b, mvs') -> loop (C.join obj a b) (mvs' @ mvs) xs)
+    in
+    loop (C.min obj) [] l
+
+  let meet (type a l) obj l =
+    let rec loop :
+        a ->
+        (a, disallowed * allowed) morphvar list ->
+        (a, l * allowed) mode list ->
+        (a, disallowed * allowed) mode =
+     fun a mvs ->
+      if C.le obj a (C.min obj)
+      then fun _ -> Amode (C.min obj)
+      else
+        function
+        | [] -> Amodemeet (a, mvs)
+        | mv :: xs -> (
+          match disallow_left mv with
+          | Amode b -> loop (C.meet obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
+          | Amodemeet (b, mvs') -> loop (C.meet obj a b) (mvs' @ mvs) xs)
+    in
+    loop (C.max obj) [] l
+
+  let zap_to_floor_morphvar ~commit obj mv =
+    let log, lower = constrain_mlower_try obj mv in
+    if commit then !append_changes log else undo_changes !log;
+    lower
+
+  let zap_to_floor : type a r. a C.obj -> (a, allowed * r) mode -> a =
+   fun obj -> function
+    | Amode a -> a
+    | Amodevar mv -> zap_to_floor_morphvar ~commit:true obj mv
+    | Amodejoin (a, mvs) ->
+      List.fold_left
+        (fun acc mv ->
+          C.join obj acc (zap_to_floor_morphvar ~commit:true obj mv))
+        a mvs
+
+  (* because lower bound conservative, this check is also conservative.
+     if it returns Some, then definitely a constant.
+     if it returns None, then we don't know anything *)
+  let check_const : type a l r. a C.obj -> (a, l * r) mode -> a option =
+   fun obj -> function
+    | Amode a -> Some a
+    | Amodevar mv ->
+      let lower = zap_to_floor_morphvar ~commit:false obj mv in
```

Comment:
This is checking to see whether the join of all the lower bounds is greater than the upper bound, right? If that is right, could we just say that directly? Then there would be no need for mutation here and no worry about committing.

Also, wouldn't that answer be exact, not conservative?

Reply:
I have said "subtree" in the other comments, but that's only conceptual. In the actual RAM, there could be circles in the graph. As we traverse the graph, we update the `lower/upper` fields of variables, and might stop if the `lower/upper` fields says there is no job to do (the first few lines in `submode_mvc/cmv` does that). In the end, the traverse traces forms a subtree.

Therefore, mutation of the `var` is crucial to know when to stop.

Reply:
Sorry -- I don't follow there. I think I'll need an example.